src/vsr/replica.zig

```diff
@@ -2220,54 +2212,126 @@ pub fn ReplicaType(
             }
             assert(self.view == message.header.view);
 
-            const view_headers = message_body_as_view_headers(message.base_const());
-            assert(view_headers.command == .start_view);
-            assert(view_headers.slice[0].op == message.header.op);
+            const view_checkpoint = start_view_message_checkpoint(message);
+            if (vsr.Checkpoint.trigger_for_checkpoint(view_checkpoint.header.op)) |trigger| {
+                assert(message.header.commit >= trigger);
+            }
+            assert(
+                message.header.op <= vsr.Checkpoint.prepare_max_for_checkpoint(
+                    vsr.Checkpoint.checkpoint_after(view_checkpoint.header.op),
+                ).?,
+            );
 
-            for (view_headers.slice) |*header| {
-                assert(header.commit <= message.header.commit);
+            const view_headers = start_view_message_headers(message);
+            assert(view_headers[0].op == message.header.op);
+            assert(view_headers[0].op >= view_headers[view_headers.len - 1].op);
+
+            if (message.header.commit > self.op_prepare_max() and (
+            //  Cluster is far ahead, replica's WAL no longer intersects primary's WAL.
+                message.header.commit > self.op_prepare_max() + constants.vsr_checkpoint_interval or
```

Comment:
Good question. I think either version is correct, and, moreover, we could completely remove this condition (we'll just hit `repair_stuck` a bit later in the future).

I'll leave this as is, though, because it seems like a nice heuristic for when to switch to state-sync proactively, and because this allows to comment on the subtelty here (which the original comment missed):

The overall condition checkifs that SV's commit is ahead of our current checkpont's prepare_max. 

**If** SV's checkpoint is our `op_checkpoint_next`, then `message.header.commit > self.op_prepare_max()` guarantees that that checkpoint is durable and is safe to sync to. 

**However**, if SV's checkpoint is further in the future, it might be the case that it is not durable. Eg, we are a severely lagging replica, primary just triggered a checkpoint and tells this fresh checkpoint to us, but the rest of the cluster is still slightly behind the cluster. But in this case it is still save to sync to that checkpoint, because our own prepares are no longer relevant. 

---

I guess if we want to tighten the heuristic where to skip the state sync, we should check which prepare we are actually missing, and sync when that gets evicted from primary's log.